lib/child_process.js

```diff
@@ -527,29 +541,55 @@ function normalizeSpawnArguments(file, args, options) {
   }
 
   if (options.shell) {
-    const command = ArrayPrototypeJoin([file, ...args], ' ');
-    // Set the shell, switches, and commands.
-    if (process.platform === 'win32') {
-      if (typeof options.shell === 'string')
-        file = options.shell;
-      else
-        file = process.env.comspec || 'cmd.exe';
-      // '/d /s /c' is used only for cmd.exe.
-      if (RegExpPrototypeTest(/^(?:.*\\)?cmd(?:\.exe)?$/i, file)) {
-        args = ['/d', '/s', '/c', `"${command}"`];
-        windowsVerbatimArguments = true;
-      } else {
-        args = ['-c', command];
-      }
+    const command = file;
+    const commandArgs = ArrayPrototypeMap(args, String);
+    args = ['-c'];
+    let quote = quotePosixShArg;
+    let isCmd = false;
+    const testWindowsExe =
+      process.platform === 'win32' ?
+        (cmd, file) =>
+          RegExpPrototypeTest(
+            new RegExp(`^(?:.*\\\\)?${cmd}(?:\\.exe)?$`, 'i'),
+            file
+          ) :
+        (_c, _f) => false;
+    // Set the shell.
+    if (typeof options.shell === 'string') {
+      file = options.shell;
+    } else if (process.platform === 'win32') {
+      file = process.env.comspec || 'cmd.exe';
+    } else if (process.platform === 'android') {
+      file = '/system/bin/sh';
     } else {
-      if (typeof options.shell === 'string')
-        file = options.shell;
-      else if (process.platform === 'android')
-        file = '/system/bin/sh';
-      else
-        file = '/bin/sh';
-      args = ['-c', command];
+      file = '/bin/sh';
+    }
+
+    // Handle special args for special shells.
+    // '/d /s /c' is used only for cmd.exe.
+    if (testWindowsExe('cmd', file)) {
+      args = ['/d', '/s', '/c'];
+      quote = quoteCmdArg;
+      isCmd = windowsVerbatimArguments = true;
+    } else if (
+      testWindowsExe('(powershell|pwsh)', file) ||
+      StringPrototypeEndsWith(file, '/pwsh')
+    ) {
+      quote = quotePwshArg;
     }
+
+    // Make it opt-in... for now.
+    if (!options.shellEscape)
+      quote = (s) => s;
+    else if (typeof options.shellEscape == 'function')
+      quote = options.shellEscape;
```

Comment:
What if `options.shellEscape` is truthy but not a function (E.G. `1`)? I think we should throw a `ERR_INVALID_CALLBACK` error in that case.

```suggestion
    if (options.shellEscape === false || options.shellEscape === undefined) {
      quote = (s) => s;
    } else if (options.shellEscape !== true) {
      validateCallback(options.shellEscape);
      quote = options.shellEscape;
    }
```

Reply:
I thought the `ERR_INVALID_ARG_TYPE` throw earliser would've caught that. When was `validateCallback` even introduced?

Reply:
> I thought the `ERR_INVALID_ARG_TYPE` throw earliser would've caught that.

Yes you're probably right. Could add a test case for this scenario please?